### PR TITLE
*: show build and probe side for join and index lookup related operators in explain result

### PR DESCRIPTION
**Pull request by LittleFall**
### What problem does this PR solve? 

Fix [14602](https://github.com/pingcap/tidb/issues/14602), and the same problem of `Apply`.

Always put Build side above Probe side.

### What is changed and how it works?

Mainly modify common_plans.go:
 - Add new argument "driverSide string" for function explainPlanInRowFormat() and prepareOperatorInfo().
 - Modify explainPlanInRowFormat()
     - Indicate driven side and driving side of parent plan.
     - Call prepareOperatorInfo() first with the Build side.
 - Modify prepareOperatorInfo(), append `driverSide` to `explainId`.

 Update test files.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has persistent data change

### CODE REVIEW

File: planner/core/common_plans.go [hunk 1]
```diff
@@ -732,19 +732,53 @@ func (e *Explain) RenderResult() error {
...
+		case *PhysicalIndexJoin:
+			buildSide = physJoinPlan.InnerChildIdx ^ 1
+		case *PhysicalIndexMergeJoin:
+			buildSide = physJoinPlan.InnerChildIdx ^ 1
+		case *PhysicalIndexHashJoin:
+			buildSide = physJoinPlan.InnerChildIdx ^ 1
+		}
+
+		if buildSide != -1 {
+			if len(driverSideInfo) < 2 {
```
Comment (francis0407) [inline planner/core/common_plans.go]: IMHO, we don't need this check. We can make sure that `PhysicalJoin` always has two children.

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -732,19 +732,53 @@ func (e *Explain) RenderResult() error {
...
+		buildSide := -1
+
+		switch physJoinPlan := physPlan.(type) {
+		case *PhysicalHashJoin:
+			if physJoinPlan.UseOuterToBuild {
+				buildSide = physJoinPlan.InnerChildIdx ^ 1
+			} else {
+				buildSide = physJoinPlan.InnerChildIdx
+			}
+		case *PhysicalMergeJoin:
```
Comment (lamxTyler) [inline planner/core/common_plans.go]: Why merge join has to differentiate the driver side?

File: cmd/explaintest/r/explain_easy_stats.result [hunk 3]
```diff
@@ -61,15 +61,15 @@ explain delete from t1 where t1.c2 = 1;
...
-  ├─IndexScan_9	0.00	cop[tikv]	table:t1, index:c2, range:[1,1], keep order:false
-  └─TableScan_10	0.00	cop[tikv]	table:t1, keep order:false
+  ├─IndexScan_9(Build)	0.00	cop[tikv]	table:t1, index:c2, range:[1,1], keep order:false
+  └─TableScan_10(Probe)	0.00	cop[tikv]	table:t1, keep order:false
 explain select count(b.c2) from t1 a, t2 b where a.c1 = b.c2 group by a.c1;
 id	count	task	operator info
 Projection_11	1985.00	root	cast(Column#8)->Column#7
 └─HashLeftJoin_21	1985.00	root	inner join, inner:HashAgg_25, equal:[eq(test.t1.c1, test.t2.c2)]
-  ├─TableReader_34	1999.00	root	data:TableScan_33
+  ├─TableReader_34(Probe)	1999.00	root	data:TableScan_33
```
Comment (SunRunAway) [inline cmd/explaintest/r/explain_easy_stats.result]: Could we always put the `Build` above the `Probe`?

File: planner/core/common_plans.go [hunk 4]
```diff
@@ -732,19 +732,53 @@ func (e *Explain) RenderResult() error {
...
 	// For every child we create a new sub-tree rooted by it.
 	childIndent := texttree.Indent4Child(indent, isLastChild)
 
 	if physPlan, ok := p.(PhysicalPlan); ok {
+		// indicate driven side and driving side of 'join'.
+		// See issue https://github.com/pingcap/tidb/issues/14602.
+		driverSideInfo := make([]string, len(physPlan.Children()))
+		buildSide := -1
+
+		switch physJoinPlan := physPlan.(type) {
```
Comment (lzmhhh123) [inline planner/core/common_plans.go]: Maybe we should also consider `apply` here?

File: cmd/explaintest/r/explain_easy_stats.result (hunk 3 above)
Comment (francis0407) [reply cmd/explaintest/r/explain_easy_stats.result]: I think the order depends on `left` or `right`?

File: planner/core/common_plans.go (hunk 1 above)
Comment (LittleFall) [reply planner/core/common_plans.go]: Thanks, I see.

File: cmd/explaintest/r/explain_easy_stats.result (hunk 3 above)
Comment (LittleFall) [reply cmd/explaintest/r/explain_easy_stats.result]: I think if one first executes `explain` as a reference, and then debugs the actual statements step by step, maybe the original format is more clear?

File: planner/core/common_plans.go (hunk 4 above)
Comment (LittleFall) [reply planner/core/common_plans.go]: Thanks, I will do it now

File: planner/core/common_plans.go (hunk 2 above)
Comment (LittleFall) [reply planner/core/common_plans.go]: "When MergeJoin is actually executed, there is a slight difference between left and right, and the outer join is more obvious."
It seems like DBAs have discussed this problem, refer to channel `team-sql-engine`.

File: cmd/explaintest/r/explain_easy_stats.result (hunk 3 above)
Comment (zz-jason) [reply cmd/explaintest/r/explain_easy_stats.result]: it's better to "always put the `Build` above the `Probe`". no matter who is on the left side.

Comment (LittleFall) [reply cmd/explaintest/r/explain_easy_stats.result]: Done.

**Comment by zz-jason** [comment]
@LittleFall please resolve the conflicts.

**Comment by LittleFall** [comment]
> LGTM.

Reply test.

**Comment by LittleFall** [comment]
> @LittleFall please resolve the conflicts.

Done. PTAL

**Pull request merged**